File: CPP/DesignMovieRentingSystem.cpp
#include <iostream>
#include <set>
#include <string>
#include <tuple>
#include <unordered_map>
#include <vector>

class MovieRentingSystem {
public:
    MovieRentingSystem(int n, std::vector<std::vector<int>>& entries) {
        (void)n;
        for (auto& e : entries) {
            int shop = e[0], movie = e[1], price = e[2];
            priceOf[key(shop, movie)] = price;
            available[movie].insert({price, shop});
        }
    }

    std::vector<int> search(int movie) {
        std::vector<int> res;
        auto it = available.find(movie);
        if (it != available.end()) {
            for (auto& [price, shop] : it->second) {
                if (res.size() == 5) break;
                res.push_back(shop);
            }
        }
        return res;
    }

    void rent(int shop, int movie) {
        int p = priceOf[key(shop, movie)];
        available[movie].erase({p, shop});
        rented.insert({p, shop, movie});
    }

    void drop(int shop, int movie) {
        int p = priceOf[key(shop, movie)];
        rented.erase({p, shop, movie});
        available[movie].insert({p, shop});
    }

    std::vector<std::vector<int>> report() {
        std::vector<std::vector<int>> res;
        for (auto& [price, shop, movie] : rented) {
            if (res.size() == 5) break;
            res.push_back({shop, movie});
        }
        return res;
    }

private:
    std::unordered_map<int, std::set<std::pair<int, int>>> available;  // movie -> {(price, shop)}
    std::set<std::tuple<int, int, int>> rented;                       // {(price, shop, movie)}
    std::unordered_map<long long, int> priceOf;

    static long long key(int shop, int movie) {
        return (long long)shop * 100000 + movie;
    }
};

std::string intList(const std::vector<int>& xs) {
    std::string s = "[";
    for (size_t i = 0; i < xs.size(); ++i) {
        if (i) s += ", ";
        s += std::to_string(xs[i]);
    }
    return s + "]";
}

std::string pairList(const std::vector<std::vector<int>>& xs) {
    std::string s = "[";
    for (size_t i = 0; i < xs.size(); ++i) {
        if (i) s += ", ";
        s += intList(xs[i]);
    }
    return s + "]";
}

int main() {
    std::vector<std::vector<int>> entries = {{0, 1, 5}, {0, 2, 6}, {0, 3, 7}, {1, 1, 4}, {1, 2, 7}, {2, 1, 5}};
    MovieRentingSystem m(3, entries);
    std::vector<std::string> out = {"null"};
    out.push_back(intList(m.search(1)));
    m.rent(0, 1);
    out.push_back("null");
    m.rent(1, 2);
    out.push_back("null");
    out.push_back(pairList(m.report()));
    m.drop(1, 2);
    out.push_back("null");
    out.push_back(intList(m.search(2)));

    std::string line = "[";
    for (size_t i = 0; i < out.size(); ++i) {
        if (i) line += ", ";
        line += out[i];
    }
    line += "]";
    std::cout << line << "\n";
    return 0;
}

File: Rust/src/bin/design_movie_renting_system.rs
use std::collections::{BTreeSet, HashMap};

struct MovieRentingSystem {
    // available[movie] = set of (price, shop); rented = set of (price, shop, movie).
    available: HashMap<i32, BTreeSet<(i32, i32)>>,
    rented: BTreeSet<(i32, i32, i32)>,
    price: HashMap<(i32, i32), i32>,
}

impl MovieRentingSystem {
    fn new(_n: i32, entries: Vec<Vec<i32>>) -> Self {
        let mut s = Self {
            available: HashMap::new(),
            rented: BTreeSet::new(),
            price: HashMap::new(),
        };
        for e in &entries {
            let (shop, movie, price) = (e[0], e[1], e[2]);
            s.price.insert((shop, movie), price);
            s.available.entry(movie).or_default().insert((price, shop));
        }
        s
    }

    fn search(&self, movie: i32) -> Vec<i32> {
        match self.available.get(&movie) {
            Some(set) => set.iter().take(5).map(|&(_, shop)| shop).collect(),
            None => Vec::new(),
        }
    }

    fn rent(&mut self, shop: i32, movie: i32) {
        let p = self.price[&(shop, movie)];
        self.available.get_mut(&movie).unwrap().remove(&(p, shop));
        self.rented.insert((p, shop, movie));
    }

    fn drop(&mut self, shop: i32, movie: i32) {
        let p = self.price[&(shop, movie)];
        self.rented.remove(&(p, shop, movie));
        self.available.get_mut(&movie).unwrap().insert((p, shop));
    }

    fn report(&self) -> Vec<Vec<i32>> {
        self.rented.iter().take(5).map(|&(_, shop, movie)| vec![shop, movie]).collect()
    }
}

fn int_list(xs: &[i32]) -> String {
    format!("[{}]", xs.iter().map(|x| x.to_string()).collect::<Vec<_>>().join(", "))
}

fn pair_list(xs: &[Vec<i32>]) -> String {
    format!("[{}]", xs.iter().map(|p| int_list(p)).collect::<Vec<_>>().join(", "))
}

fn main() {
    let mut m = MovieRentingSystem::new(
        3,
        vec![vec![0, 1, 5], vec![0, 2, 6], vec![0, 3, 7], vec![1, 1, 4], vec![1, 2, 7], vec![2, 1, 5]],
    );
    let mut out = vec!["null".to_string()];
    out.push(int_list(&m.search(1)));
    m.rent(0, 1);
    out.push("null".into());
    m.rent(1, 2);
    out.push("null".into());
    out.push(pair_list(&m.report()));
    m.drop(1, 2);
    out.push("null".into());
    out.push(int_list(&m.search(2)));
    println!("[{}]", out.join(", "));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example() {
        let mut m = MovieRentingSystem::new(
            3,
            vec![vec![0, 1, 5], vec![0, 2, 6], vec![0, 3, 7], vec![1, 1, 4], vec![1, 2, 7], vec![2, 1, 5]],
        );
        assert_eq!(m.search(1), vec![1, 0, 2]);
        m.rent(0, 1);
        m.rent(1, 2);
        assert_eq!(m.report(), vec![vec![0, 1], vec![1, 2]]);
        m.drop(1, 2);
        assert_eq!(m.search(2), vec![0, 1]);
    }
}
